fix(codecs) csv documentation defaults as ascii_char

scMarkus opened the pull request:
<!--
**Your PR title must conform to the conventional commit spec!**

  <type>(<scope>)!: <description>

  * `type` = chore, enhancement, feat, fix, docs
  * `!` = OPTIONAL: signals a breaking change
  * `scope` = Optional when `type` is "chore" or "docs", available scopes https://github.com/vectordotdev/vector/blob/master/.github/semantic.yml#L20
  * `description` = short description of the change

Examples:

  * enhancement(file source): Add `sort` option to sort discovered files
  * feat(new source): Initial `statsd` source
  * fix(file source): Fix a bug discovering new files
  * chore(external docs): Clarify `batch_size` option
-->

work in progress PR to work out how to best solve #18669.

current state has been described here https://github.com/vectordotdev/vector/issues/18669#issuecomment-2110454821

@tobz woudl you suggest looking into the conversion between json schema file and cue documentation? if so may I ask you for a hint on where best to start?

scMarkus commented:
@tobz sorry for taking so long. I am now able to focus on this issue again. 

From what I understand the rust specific part of adding meta data to a type by using mcros like `    #[configurable(metadata(docs::type_override = "ascii_char"))]` seams to work reasonable. the generated json output does contain all of the needed information. What I think is missing, would be a correct handling of such information in the ruby script generating cue files. I have added a naive code branch handling my specific case. Pleas have a look at https://github.com/vectordotdev/vector/pull/20498/commits/d49d23985332700dda26c5de4bae34a0e252742e. The output does look like what I would expect in the cue files but I have not validated it in any rendered form since I would not know how.

I am quite certain that this is not the final solution we want to have so I am curious about your thoughts @tobz.

drichards-87 commented on website/content/en/highlights/2024-06-17-0-39-0-upgrade-guide.md:
```diff
@@ -0,0 +1,42 @@
+
+#### Path coalescing has been removed {#path-coalescing}
+
+Coalescing of field lookup paths (e.g. `.(field1|field2)`) was deprecated and is now removed This
```
```suggestion
Coalescing of field lookup paths (e.g., `.(field1|field2)`) was deprecated and is now removed. This
```

drichards-87 commented on website/content/en/highlights/2024-06-17-0-39-0-upgrade-guide.md:
```diff
@@ -0,0 +1,42 @@
+Coalescing of field lookup paths (e.g. `.(field1|field2)`) was deprecated and is now removed This
+feature did not seem to be used much and significantly complicates parts of the codebase.
+
+If you were using this feature, you can accomplish the same through conditionals in VRL like:
```
```suggestion
If you were using this feature, you can accomplish the same through conditionals in VRL:
```

drichards-87 commented on website/content/en/highlights/2024-06-17-0-39-0-upgrade-guide.md:
```diff
@@ -0,0 +1,42 @@
+
+#### Removal of `enterprise` configuration {#enterprise-configuration}
+
+The `enterprise` global configuration option that was used to integrate Vector with the Datadog has been removed.
```
```suggestion
The `enterprise` global configuration option that was used to integrate Vector with Datadog has been removed.
```

drichards-87 commented on website/content/en/highlights/2024-06-17-0-39-0-upgrade-guide.md:
```diff
@@ -0,0 +1,42 @@
+#### Removal of `enterprise` configuration {#enterprise-configuration}
+
+The `enterprise` global configuration option that was used to integrate Vector with the Datadog has been removed.
+Observability Pipelines product was deprecated and has been removed. Please reach out to Datadog
```
```suggestion
The Observability Pipelines product was deprecated and has been removed. Please reach out to Datadog
```

drichards-87 replied:
It might be useful to include a link/method for contacting Datadog support.

drichards-87 reviewed (approved):
Left some small suggestions from Docs and approved the PR.

tobz commented:
@scMarkus Nice, the changes look good to me.

I thought we used to have a CI job that would publish a preview version of the docs based on the PR branch, but it seems like not... let me figure out the best way to preview the fully-generated docs and get back to you.

tobz commented:
So, looking at the HTML template that gets used to render this part of the configuration docs, I'm pretty confident this should just work as-is.

It will end up looking like this:

![Screenshot 2024-06-27 at 10 52 28 AM](https://github.com/vectordotdev/vector/assets/222615/ae17152a-0cb1-4715-b827-c10823c2d410)

While `ascii_char` is maybe a little weird, I don't think we need to worry about that here. The end result it's going to properly show a comma (or whatever the character is) and that seems like a win to me.

scMarkus commented:
I have added the other two configuration parameters to our new `ascii_char` handling and added a note in the ruby script that the new behavior is meant for u8 data type only. From my point of view this would solve the issue as described in #18669.

tobz commented on scripts/generate-component-docs.rb:
```diff
@@ -744,9 +744,19 @@ def resolve_schema(root_schema, schema)
   # We intentially set no actual definition for these types, relying on the documentation generation
   # process to provide the actual details. We only need to specify the custom type name.
+  #
+  # To handle u8 types as ascii characters and not there uint representation between 0 and 255 we
+  # added a special handling of these exact values. This means
+  # `#[configurable(metadata(docs::type_override = "ascii_char"))]` should only be used consciously
+  # for rust u8 type. See lib/codecs/src/encoding/format/csv.rs for an example and
+  # https://github.com/vectordotdev/vector/pull/20498
```
```suggestion
# For `u8` fields that have a type override of `ascii_char` type, we specifically use the
# character representation of the default value, which ensures that we display the default
# values as an actual string character instead of an ordinal value.
#
# See `lib/codecs/src/encoding/format/csv.rs` for an example.
```

tobz reviewed (changes requested):
Just one small suggested changed in the doc comment for the generation script.

scMarkus commented:
In regards to the latest failed test run. if I understand the [output](https://github.com/vectordotdev/vector/actions/runs/9730431193/job/26862790199?pr=20498) correctly, there is an issue with `make check-docs`. Running this command locally does not find such errors. Pleas let me know if I can change anything to make the tests pass.

local execution:
```bash
~/tools/vector $ make check-docs ENVIRONMENT=true CONTAINER_TOOL="podman"
Building the environment. (ENVIRONMENT_AUTOBUILD=true) This may take a few minutes...
podman build --quiet --tag docker.io/timberio/vector-dev:sha-3eadc96742a33754a5859203b58249f6a806972a --file scripts/environment/Dockerfile .
6a585b575749b9677fc886abdaeadb26a80751da2e29e2a7f1f764e2c81be7ef
Entering environment...
podman run --name vector-environment --rm --tty --init --interactive --env INSIDE_ENVIRONMENT=true --network host --mount type=bind,source=/home/markus/tools/vector,target=/git/vectordotdev/vector  --mount type=volume,source=vector-target,target=/git/vectordotdev/vector/target --mount type=volume,source=vector-cargo-cache,target=/root/.cargo --mount type=volume,source=vector-rustup-cache,target=/root/.rustup  docker.io/timberio/vector-dev:sha-3eadc96742a33754a5859203b58249f6a806972a cargo vdev check docs
Skipping cue files format validation - reserved for CI
Validating cue files correctness...
Success! The contents of the sources in the "./website/cue" directory are valid
```

scMarkus commented:
Thus far I feel positive about this pull request. @tobz you have approved it already as well. Is there anything left to do in your opinion?

tobz commented:
Sorry, this fell off my radar.

I agree that this is ready to go. 👍🏻

scMarkus commented:
No worries. Thanks for having this progressing :)
